#12 Place an order

mayerij opened the pull request:
place an order (or a list of orders) ;
shopOrderId is back
Sender -> MessageService (and kafka configuration comes  from properties)

mayerij committed:
bonnie-12 Shop - Place an order

mayerij committed:
bonnie-12 Shop - Place an order main merge 1st step

mayerij committed:
bonnie-12 Shop - Place an order main merge 2st steps
add shopId

mayerij committed:
bonnie-12 Shop - Place an order main merge 2st steps
add shopId
fix shopId

mayerij committed:
bonnie-12 Shop - Place an order main merge 3rd step ;
add shopId;
fix shopId;
log.info -> log debug;
fixing modified tests;

abalazsik commented on communicationPlugIn-Kafka/src/main/java/com/cgi/hexagon/commplugin/kafka/OrderConsumer.java:
```diff
@@ -0,0 +1,51 @@
+        log.debug("Message received in Group: [{}] Topic: [{}] Message: [{}]", groupID, topicName, message);
+        List<OrderJson> jsonOrders;
+        if (message.trim().startsWith("[")) //Json array
+            jsonOrders = new JsonMapper().readAll(message);
```
This JsonMapper should be a field, so we won't recreate it on every call

mayerij replied:
It's the question of CPU time vs. memory allocation
it happens very rearly ... but I change
..
same too all "field" comments

abalazsik commented on communicationPlugIn-Kafka/src/main/java/com/cgi/hexagon/commplugin/kafka/OrderConsumer.java:
```diff
@@ -0,0 +1,51 @@
+        if (message.trim().startsWith("[")) //Json array
+            jsonOrders = new JsonMapper().readAll(message);
+        else
+            jsonOrders = Collections.singletonList(new JsonMapper().read(message));
```
This too

mayerij replied:
same

abalazsik commented on communicationPlugIn-Kafka/src/main/java/com/cgi/hexagon/commplugin/kafka/OrderConsumer.java:
```diff
@@ -0,0 +1,51 @@
+        jsonOrders = jsonOrders.stream().filter(Objects::nonNull).collect(Collectors.toList());
+
+        if (jsonOrders.size() > 0)
+            orderService.createOrders(new OrderMapper().fromOrderJsonList(jsonOrders));
```
This OrderMapper can be a field too

abalazsik commented on h2-storage/pom.xml:
```diff
@@ -34,6 +34,11 @@
+		<groupId>org.json</groupId>
+		<artifactId>json</artifactId>
+		<version>20220320</version>
+	</dependency>
```
Do we need this?

abalazsik commented on h2-storage/src/main/java/com/cgi/hexagon/h2storage/order/H2OrderStorage.java:
```diff
@@ -52,15 +60,22 @@ public List<Order> findAll() {
+   
+    public List<Order> findAllByShopOrderId(String shopOrderId) {
+        List<Order> result = new ArrayList<>();
+        orderRepository.findAllByShopOrderId(shopOrderId).forEach(order -> result.add(mapper.fromEntity(order)));
```
return orderRepository.findAllByShopOrderId(shopOrderId).stream().map(mapper::fromEntity).collect(Collectors.toList());

mayerij replied:
of course :-)   the parameter was Iterable not Collection by the source of this copied method

abalazsik commented on communicationPlugIn-Kafka/src/main/java/com/cgi/hexagon/commplugin/kafka/OrderMapper.java:
```diff
@@ -0,0 +1,23 @@
+    public List<Order> fromOrderJsonList(List<OrderJson> orderJsons) {
+        List<Order> ret = new ArrayList<Order>();
+        for (OrderJson orderJson : orderJsons) {
+            ret.add(fromOrderJson(orderJson));
```
return orderJsons.stream().map(this::fromOrderJson).collect(Collectors.toList());

abalazsik commented on bonnie-businessrules/src/main/java/com/cgi/hexagon/businessrules/order/OrderService.java:
```diff
@@ -75,29 +80,72 @@ public boolean setTrackingNumber(long id, String trackingNr) {
+        }
+        if (order.getQuantity() < 1) {
+            log.error(" Invalid quantity in {}", order.toString());
+            return -1;
```
If we return with error codes, maybe it is not a bad idea the have distinguishable return codes (-2, -3...)

abalazsik commented on bonnie-businessrules/src/test/java/com/cgi/hexagon/businessrules/OrderServiceTest.java:
```diff
@@ -226,18 +228,24 @@ public void expectSetTrackingNumberReturnsTrue() {
+        public void expectSetTrackingNumberCallsSender() {
+
+           when(orderLoader.load(ORDER_ID)).thenReturn(getOrder().withStatus(Status.ASSEMBLED));
+       when( sender.send( any())).thenReturn( true);
```
Should be something like: doNothing().when(sender.send(any()))

abalazsik commented on communicationPlugIn-Kafka/src/main/java/com/cgi/hexagon/commplugin/kafka/JsonMapper.java:
```diff
@@ -0,0 +1,44 @@
+public class JsonMapper {
+
+    public OrderJson read(String jsonString) {
+        final ObjectMapper objectMapper = new ObjectMapper();
```
This ObjectMapper should be a field, so we won't recreate it on every call

abalazsik commented on communicationPlugIn-Kafka/src/main/java/com/cgi/hexagon/commplugin/kafka/JsonMapper.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public List<OrderJson> readAll(String jsonString) {
+        final ObjectMapper objectMapper = new ObjectMapper();
```
This too

abalazsik commented on communicationPlugIn-Kafka/src/main/java/com/cgi/hexagon/commplugin/kafka/JsonMapper.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public String write(SendRequest sendRequest) {
+        final ObjectMapper objectMapper = new ObjectMapper();
```
This too

abalazsik commented on communicationPlugIn-Kafka/src/main/java/com/cgi/hexagon/commplugin/kafka/KafkaMessageService.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    public void send(SendRequest request) {
+        kafkaTemplate.send(topicName, new JsonMapper().write(request));
```
This JsonMapper can be a field

mayerij committed:
bonnie-12 Shop - Place an order main merge 4th step ;
add shopId;
fix shopId;
log.info -> log debug;
fixing modified tests;
fixing Adorján's commits

KenyeresKornel commented on communicationPlugIn-Kafka/src/main/java/com/cgi/hexagon/commplugin/kafka/OrderJson.java:
```diff
@@ -0,0 +1,67 @@
+    String goodsId;
+    int quantity;
+
+    JsonNode metadata;
```
add placementDate property as well please. Based on our last discussion this is not a parameter that we want to set on realization, the shop is going to send this as well

mayerij committed:
bonnie-12 Shop - Place an order main merge 4th step ;
add shopId;
fix shopId;
log.info -> log debug;
fixing modified tests;
fixing Adorján's commits;
placementDate added;
some comment (I hope all) fixed

petr-vanis-cgi commented:
has this been approved for merging by someone already?

petr-vanis-cgi commented:
ok going forward I would like to only merge things that have formal approval in the pull request itself so that I can be also angry with people not doing correct reviews ;-) 
in my opinion this one should have not yet been merged
